src/index.ts

```diff
@@ -0,0 +1,409 @@
+import { join } from 'path';
+import { readFile, readdir, rename, appendFile, writeFile, mkdir } from 'fs/promises';
+import config from 'config';
+import { Tracing } from '@map-colonies/telemetry';
+import { trace as traceAPI, context as contextAPI, SpanStatusCode, SpanKind, SpanStatus, Span } from '@opentelemetry/api';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import execa from 'execa';
+import { ObjectCannedACL } from '@aws-sdk/client-s3';
+import {
+  BACKUP_DIR_NAME,
+  DIFF_FILE_EXTENTION,
+  ExitCodes,
+  OsmdbtCommand,
+  OSMDBT_BIN_PATH,
+  OSMDBT_CONFIG_PATH,
+  OSMDBT_DONE_LOG_PREFIX,
+  S3_LOCK_FILE_NAME,
+  STATE_FILE,
+} from './constants';
+import { logger } from './telemetry/logger';
+import { OsmdbtConfig, ObjectStorageConfig, TracingConfig } from './interfaces';
+import { ErrorWithExitCode } from './errors';
+import { getDiffDirPathComponents, streamToString } from './util';
+import { FsSpanName, FsAttributes } from './telemetry/tracing/fs';
+import { OsmdbtAttributes, OsmdbtSpanName } from './telemetry/tracing/osmdbt';
+import { JobAttributes, ROOT_JOB_SPAN_NAME } from './telemetry/tracing/job';
+import { handleSpanOnSuccess, handleSpanOnError, promisifySpan, TRACER_NAME } from './telemetry/tracing/util';
+import { deleteObjectWrapper, getObjectWrapper, headObjectWrapper, putObjectWrapper } from './s3';
+
+const tracing = new Tracing();
+tracing.start();
+const tracer = traceAPI.getTracer(TRACER_NAME);
+const rootJobSpan = tracer.startSpan(ROOT_JOB_SPAN_NAME, { attributes: { [JobAttributes.JOB_ROLLBACK]: false } });
+
+const objectStorageConfig = config.get<ObjectStorageConfig>('objectStorage');
+const tracingConfig = config.get<TracingConfig>('telemetry.tracing');
+const osmdbtConfig = config.get<OsmdbtConfig>('osmdbt');
+
+const OSMDBT_STATE_PATH = join(osmdbtConfig.changesDir, STATE_FILE);
+const OSMDBT_STATE_BACKUP_PATH = join(osmdbtConfig.changesDir, BACKUP_DIR_NAME, STATE_FILE);
+const GLOBAL_OSMDBT_ARGS = osmdbtConfig.verbose ? ['-c', OSMDBT_CONFIG_PATH] : ['-c', OSMDBT_CONFIG_PATH, '-q'];
+
+let jobExitCode = ExitCodes.SUCCESS;
+let isS3Locked = false;
+let filesUploaded = 0;
+
+for (const signal of ['SIGTERM', 'SIGINT']) {
+  process.on(signal, () => {
+    cleanup(ExitCodes.TERMINATED).finally(() => process.exit(ExitCodes.TERMINATED));
+  });
+}
+
+const prepareEnvironment = async (span?: Span): Promise<void> => {
+  const { logDir, changesDir, runDir } = osmdbtConfig;
+
+  logger.debug({ msg: 'preparing environment', osmdbtDirs: { logDir, changesDir, runDir } });
+
+  const backupDir = join(changesDir, BACKUP_DIR_NAME);
+  const uniqueDirs = [logDir, changesDir, runDir, backupDir].filter((dir, index, dirs) => dirs.indexOf(dir) === index);
```

Comment:
Can you please explain what is the purpose of this line?

Reply:
`logDir`, `changesDir`, `runDir` are retrieved from `osmdbt-config.yaml` (see `osmdbt-config-example.yaml`), these values might overlap. `backupDir` is retrieved according to `changesDir` value and is used in the case of a rollback.
`uniqueDirs` contains these 4 dirs with no overlaps and is used to `mkdir` each of them before the business logic is initialized